### DISCUSSION TITLE
What is the state of the joomla framework?

### DISCUSSION THREAD

**Discussion by dmitriitux** [joomla/joomla-cms#36175] [Q&A]
Hello everybody!
My task is to create a microservice, I will build it in php. I need base classes for this which the joomla framework has. Without cms structure and extensions for it.
The question is, what is the state of the framework? Can it be used in a combat project?

**Comment by dmitriitux** [comment]
@brianteeman Hi, can you answer this question? or do you know who can give the answer?

**Comment by HLeithner** [comment]
Atm main maintainer of the framework is @nibra 

Project by Joomla using the framework 2.0
https://github.com/joomla/help.joomla.org
and
https://github.com/joomla/framework.joomla.org

hope that helps you start

**Reply by dmitriitux** [reply]
thanks, Where are these repositories used? 
https://github.com/orgs/joomla-framework/repositories

**Reply by HLeithner** [reply]
installed with composer https://github.com/joomla/help.joomla.org/blob/master/composer.json ?

**Reply by dmitriitux** [reply]
and, I realized, these are just site repositories =)

**Comment by dmitriitux** [comment]
@nibra hi, can you answer the question I asked here above? about the stability of the framework, we need to load some of the joomla components with a composer, to work with the database, files, cache and etc.

**Reply by HLeithner** [reply]
since the cms uses the framework it's pretty stable

**Reply by dmitriitux** [reply]
that is, we can safely use separate repositories and download through composer?

**Reply by HLeithner** [reply]
that's the same way as the cms uses it so yes

**Reply by dmitriitux** [reply]
https://github.com/joomla/joomla-cms/tree/4.1-dev/libraries/src/
https://github.com/orgs/joomla-framework/repositories

and here's another moment. The framework is expanding in cms, so errors can be fixed in cms, that is, the framework itself, separately from cms, may contain errors and is probably not tested separately?

And yet, as an example of the Input class, consider, in the framework there are purely abstractions and the implementation is precisely in the overrides of the Input class in cms, that is, the framework itself is sometimes useless without implementation.

**Reply by dmitriitux** [reply]
@HLeithner Why am I, if the framework is not tested separately, then it is not stable, because if we update from the composer, this is all potentially unstable and we need to cover the entire microservice with unit tests

**Reply by HLeithner** [reply]
The framework has it own unit tests and should be tested very well (better then the cms).

Sure the cms has special requirements that can't be provided by the framework because the repositories do only one thing and (normally) don't interact or require another framework component.

So basically if you like to use the framework you use parts of it but don't build the cms with it (that doesn't make sense because the cms is already here).

For example I use the database layer, the inputfilter and sometimes the registry for my projects that don't use the cms.

**Reply by nibra** [reply]
@dmitriitux The framework is stable and production ready. It is heavily tested. If you're not convinced, clone the package you want to use and run the included tests. I use the framework in a lot of projects myself.

**Reply by dmitriitux** [reply]
making your own cms is pointless, that's understandable.
I understood about testing.

What about the implementation? then that in the framework it is purely abstraction sometimes, and the implementation in cms? it turns out you need to write your own implementation in the same way, for example, processing the same Input

**Reply by nibra** [reply]
> that in the framework it is purely abstraction sometimes

What do you mean?

**Reply by dmitriitux** [reply]
@nibra 
I meant that the implementation can be in cms, but in the framework it is not there and you have to write it yourself. but it was I who communicated with colleagues who tried to use the framework and encountered this. Now I rummaged myself and see that the current code in the overrides in cms and frameworks differ slightly. I will study in more detail and will report here if I notice this, there are strong differences between the framework and cms.

Thank you for your answers.

**Comment by wilsonge** [comment]
For the CMS we override Input specifically as an example purely because of the backwards compatible implications for typehinting. As you can see we've deprecated all the classes in favour of the framework classes and there are no major differences. In 4.x we changed the inheritance tree to extend from the framework to allow for the typehints to be against the framework class and in 5.0 we'll remove the CMS classes entirely.

There are also concrete cases (e.g. filter) where there might be cms specific implementations - such as the username regex's etc. but these things are class specific.
